src/ddmd/dmangle.d

```diff
@@ -385,10 +479,14 @@ public:
         if (p)
         {
             mangleParent(p);
-            if (p.getIdent())
+            auto ti = p.isTemplateInstance();
+            if (ti && !ti.isTemplateMixin())
+            {
+                mangleTemplateInstance(ti);
+            }
+            else if (p.getIdent())
```

Comment:
This change caused a regression (or exposed an underlying bug).

Type merging of `immutable!T` -> `T` is now broken, if `T` is nested in a `struct S(alias fun)`.  Calling mangleToBuffer is called before and after `fun` has completed its semantic pass now yields two different results.  So there are two copies of `T` floating around the AST with pointers to `immutable!T`.

```
struct MultiwayMerge()
{
    bool compFront()
    {
        return true;
    }

    struct BinaryHeap(alias less)
    {
        struct RefCounted
        {
            struct Impl
            {
                int _payload;
            }

            void initialize()
            {
                immutable Impl init;
            }
        }
    }

    BinaryHeap!(compFront) _heap;
}

MultiwayMerge!() multiwayMerge()
{
    return typeof(return)();
}
```
```

Reply:
> Calling mangleToBuffer is called before and after fun has completed its semantic pass now yields two different results.

AFAICT calling mangleToBuffer on an incompletely analyzed symbol should never happen. The old implementation reused the first (incorrect) result, so the issue could have passed unnoticed.